docs/ImportONNXDefs.md

```diff
@@ -1,46 +1,89 @@
+'version_dict', in gen_onnx_mlir.py. 
+The key of this directory is the operation name and the value is the list of 
+opset for this operation. Usually only the top version opset of this operation (in onnx-mlir/third_party/onnx) is supported. Details about versioning can be found in [version section](#Operation Version).
+With this entry, the script will generate the operation defintion for onnx dialect.
 
-ONNX specifications are defined under `onnx/defs` directory in the ONNX project repository. 
-There is a python script [utils/gen_onnx_mlir.py](../utils/gen_onnx_mlir.py) that automatically generates documents about operations in ONNX (docs/Operations.md). 
-ONNX-MLIR modified this script to import ONNX specifications into ONNX-MLIR. 
-There are two files generated for ONNX MLIR with the modified gen_onnx_mlir.py:
+# Customization <a name="customize"></a>
 
-1. `src/Dialect/ONNX/ONNXOps.td.inc`: Operation definition for MLIR TableGen. `src/Dialect/ONNX/ONNXOps.td` includes this file.
-2. `src/Builder/OpBuildTable.inc`: C++ code for ONNX-MLIR frontend to import operation nodes from ONNX model. `src/Builder/FrontendDialectTransformer.cpp` includes this file.
+## Add Interface and Trait
+By default, all operation has shape inference interface and `NoSideEffect` trait.
+If an opration has `ResultTypeInferenceOpInterface`, use dictionary `OpsWithResultTypeInference`. 
+This interface inferes the type of result tensor, not shape. 
+If an operation has subgraph, it will has interface `HasOnnxSubgraphOpInterface`. 
 
-## How to use the script
-1. Install [ONNX](https://github.com/onnx/onnx). We highly recommend that you use the one located at `third_party/onnx.` On many systems, it can be installed by executing the `pip install -e third_party/onnx` command in the top onnx-mlir directory.
-2. Make target `OMONNXOpsIncTranslation`. For example, run the command below in your onnx-mlir build directory.
+## Add canonicalization interface
+If a transformation should be applied locally to an operation across passes, canonicalization 
+interface can be used for this transformation. To enable the canonicalization for an operation, 
+add the name of this operation into this list of  `OpsWithCanonicalizer` and then the operation 
+will have `hasCanonicalizer = 1;` in its definition.
+
+## Customize builder
+The default builders for an operation require the type of results as a parameter. However, the type
+of results can be inferred. A customize builder may be a useful to simplify the code. Based on the
+type of inference, there are two kinds builder, unranked type and broadcast type. To enable the 
+special builder for an operation, you can add its name into `custom_builder_unranked_ops_list`
+ and `custom_builder_broadcast_ops_list` respectively.
+
+Please note that the need of special builder in rewriting rules can be avoided with the use of `returnType`. It may be a better solution to just move such type inference code into ONNXOpHelper.cpp
```

Comment:
Its probably because I have not used returnType, it may help to provide a bit more info. And maybe if there is an operation that does this well, say "look at op X for a good example of how this is done".

Reply:
I will add an example. Recently, Tung tried "returnType" in his new PR and got rid of the customized builder definition. We may only need to keep the broadcast builder. I will record this as a future work.